Allow template for the payload parameter for 'rest sensor'

jantenhove opened the pull request:
## Description:

Allow templates for the payload parameter.
This is needed for some webservices, like the APS inverters that need to include the current date in the payload. For example:
```yaml
- platform: rest 
  name: "APS opgewekt vandaag"
  method: POST
  resource: "https://apsystemsema.com/ema/ajax/getDashBoardAjax/getSystemCurrentDayEnergy"
  headers:
    User-Agent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36"
    Content-Type: "application/x-www-form-urlencoded"
    Accept: "application/json, text/javascript, */*; q=0.01"
    X-Requested-With: XMLHttpRequest
    Origin: https://apsystemsema.com
    Referer: https://apsystemsema.com/ema/security/optmainmenu/intoLargeDashboard.action
  payload: "eculist=21600000xxx&date={{ as_timestamp(now()) | timestamp_custom('%Y%m%d')}}"
  value_template: "{{(value | float) | round(2)}}"
  device_class: power
  unit_of_measurement: 'kWh'
  scan_interval: 300
  force_update: true
  timeout: 30
```

## Checklist:
  - [ x] The code change is tested and works locally.
  - [ x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [ x] There is no commented out code in this PR.
  - [ x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [ x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)


[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

jantenhove committed:
Allow template for payload

Allow templates for the payload parameter

jantenhove committed:
Update payload every time sensor updates

Updates the payload with template on each sensor update

jantenhove committed:
Remove logging of request body

homeassistant commented:
Hi @jantenhove, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#25417).

Once you do that we will be able to review and accept this pull request.

Thanks!

frenck commented:
This causes every payload to be handled as a template, wouldn't it be more viable to introduce a `payload_template` option?

pvizeli commented:
Please address CI issues

jantenhove commented:
There is no information on the CI issues besides: "Bash exited with code '1'." I don't have a clue how to solve this.

MartinHjelmare commented:
Solve the merge conflict first. Then the build will run again and we can see what's wrong.

jantenhove committed:
Merge branch 'dev' into jantenhove-patch-rest-payload

jantenhove commented:
Tip: Solving merge conflicts with Github is not a good idea :). I will fix the Lint issues and commit again.

jantenhove committed:
Fix lint issues

jantenhove committed:
Black reformat

jantenhove commented:
Why is CheckFormat failing? I've used the Black formatter to format the code. The error message is not very helpful:
```
in/bash --noprofile --norc /__w/_temp/d7fbf2b5-16a6-4243-838b-99cb013d66a2.sh
##[error]Bash exited with code '1'.
Finishing: Check Black formatting

MartinHjelmare commented:
Did you run black from the project root and commit the changes?
```
black --fast homeassistant
```

jantenhove committed:
Black reformat

Reformatted using black (new try)

jantenhove committed:
Binary sensor supports template for payload

Binary Rest sensor depends on RestData declared in sensor.py which now requires a template instead of a string.

jantenhove commented:
Do I need to do more to get this PR merged?

MartinHjelmare commented:
Please solve the merge conflict.

springstan commented:
@jantenhove any updates on this? :)

balloob commented:
This PR seems to have gone stale. Closing it.
